domain/ide/workspace/folder.go

```diff
@@ -79,21 +79,58 @@ type Folder struct {
 	notifier                noti.Notifier
 }
 
-func (f *Folder) Issues() map[string][]snyk.Issue {
+func (f *Folder) ClearIssues(path string) {
+	if cacheProvider, isCacheProvider := f.scanner.(snyk.CacheProvider); isCacheProvider {
+		cacheProvider.ClearIssues(path)
+	}
+}
+
+func (f *Folder) RegisterCacheRemovalHandler(handler func(path string)) {
+	if cacheProvider, isCacheProvider := f.scanner.(snyk.CacheProvider); isCacheProvider {
+		cacheProvider.RegisterCacheRemovalHandler(handler)
+	}
+}
+
+func (f *Folder) IssuesByProduct() snyk.ProductIssuesByFile {
+	issuesForProduct := snyk.ProductIssuesByFile{}
+	for path, issues := range f.Issues() {
+		if !f.Contains(path) {
+			panic("issue found in cache that does not pertain to folder")
+		}
+		for _, issue := range issues {
+			p := issue.Product
+			productIssuesByFile := issuesForProduct[p]
+			if productIssuesByFile == nil {
+				productIssuesByFile = snyk.IssuesByFile{}
+			}
+			productIssuesByFile[path] = append(productIssuesByFile[path], issue)
+			issuesForProduct[p] = productIssuesByFile
+		}
+	}
+	return issuesForProduct
+}
+
+func (f *Folder) Issues() snyk.IssuesByFile {
 	// we want both global issues (OSS and IaC at the moment) and scanner-local issues (Code at the moment)
 	// so we get the global issues first, then append the scanner-local issues
-	issues := make(map[string][]snyk.Issue)
-	f.documentDiagnosticCache.Range(func(key string, value []snyk.Issue) bool {
-		issues[key] = value
+	issues := snyk.IssuesByFile{}
+	f.documentDiagnosticCache.Range(func(path string, value []snyk.Issue) bool {
+		if f.Contains(path) {
+			issues[path] = value
+		} else {
+			panic("issue found in cache that does not pertain to folder")
```

Comment:
Same here, we should not have issues in our global folder cache that do not belong to the folder.